[OpenMP][CodeExtractor]Add align metadata to load instructions

DominikAdamski opened the pull request:
Moving code to another function can lead to missed optimization opportunities, because function passes operate on smaller chunks of code, and they cannot figure out all details.

One example of missed optimization opportunities after code extraction is information about pointer alignment. The instruction combine pass adds information about pointer alignment to LLVM intrinsic memcpy calls if it can deduce it from the code or if align metadata is added. If this information is not present, then further optimization passes can generate inefficient code.

If we add align metadata to extracted pointers, then the instruction combine pass can add the align attribute to the LLVM intrinsic memcpy call and unblock further optimization.

Scope of changes:
1. Analyze MLIR map operations. Add information about the alignment of objects that are passed by reference to OpenMP GPU kernels.
2. Propagate alignment information to the outlined by `CodeExtractor` helper functions.

DominikAdamski committed:
[CodeExtractor] Add align metadata to extracted pointers

Moving code to another function can lead to missed optimization opportunities,
because function passes operate on smaller chunks of code,
and they cannot figure out all details.

One example of missed optimization opportunities after code extraction
is information about pointer alignment. The instruction combine pass
adds information about pointer alignment to LLVM intrinsic memcpy calls
if it can deduce it from the code or if align metadata is added.
If this information is not present, then further optimization passes
can generate inefficient code.

If we add align metadata to extracted pointers, then the instruction
combine pass can add the align attribute to the LLVM intrinsic memcpy call
and unblock further optimization.

Meinersbur commented on llvm/lib/Transforms/Utils/CodeExtractor.cpp:
```diff
@@ -1604,8 +1604,26 @@ void CodeExtractor::emitFunctionBody(
+                       "loadgep_" + inputs[i]->getName(), newFuncRoot);
+      PointerType *ItemType =
+          dyn_cast<PointerType>(StructArgTy->getElementType(aggIdx));
+      if (ItemType && !LoadGEP->getMetadata(LLVMContext::MD_align)) {
```
Is the `!LoadGEP->getMetadata()` condition superfluous? How can there be metadata if we just created the instruction?

DominikAdamski replied:
Fixed, thanks for pointing out.

Meinersbur reviewed (commented):
I had some questions about alignment:
 
 * The StructArgTy is allocated at https://github.com/llvm/llvm-project/blob/e3c80d4496d109397e3c6403df434513fcd50bc6/llvm/lib/Transforms/Utils/CodeExtractor.cpp#L1809. Does this guaranteed to have the alignment returned by `getPointerPrefAlignment`? If so, could add a comment?

 * ~~Shouldn't there be a pass or similar that propagates alignment information from the alloca to the load? Interprocedurally we have the attributor, so what is doing this within a function?~~ It's not the same function, the LoadGEP is in the extracted function.

DominikAdamski committed:
Applied remarks

DominikAdamski commented:
> I had some questions about alignment:
> 
> * The StructArgTy is allocated at https://github.com/llvm/llvm-project/blob/e3c80d4496d109397e3c6403df434513fcd50bc6/llvm/lib/Transforms/Utils/CodeExtractor.cpp#L1809
>   . Does this guaranteed to have the alignment returned by `getPointerPrefAlignment`? If so, could add a comment?
> * ~Shouldn't there be a pass or similar that propagates alignment information from the alloca to the load? Interprocedurally we have the attributor, so what is doing this within a function?~ It's not the same function, the LoadGEP is in the extracted function.

I added a comment and ensured that we use the same function for alignment calculation. I modified the struct allocation - I explicitly set the alignment for struct allocation. Changes in the struct allocation constructors are non-functional. I explicitly set the alignment parameter in the constructor just to ease code analysis.

Previous constructor:
https://github.com/llvm/llvm-project/blob/0a5847f1c1445bfe55c9f9241c4d6d89609a7dc0/llvm/lib/IR/Instructions.cpp#L1260
calls `DL.getPrefTypeAlign`:
https://github.com/llvm/llvm-project/blob/0a5847f1c1445bfe55c9f9241c4d6d89609a7dc0/llvm/lib/IR/Instructions.cpp#L1246

DominikAdamski committed:
Merge branch 'main' into codeextractor_add_align_metadata

Meinersbur commented on llvm/lib/Transforms/Utils/CodeExtractor.cpp:
```diff
@@ -1604,8 +1604,26 @@ void CodeExtractor::emitFunctionBody(
+      LoadInst *LoadGEP =
+          new LoadInst(StructArgTy->getElementType(aggIdx), GEP,
+                       "loadgep_" + inputs[i]->getName(), newFuncRoot);
+      PointerType *ItemType =
+          dyn_cast<PointerType>(StructArgTy->getElementType(aggIdx));
+      if (ItemType) {
```
```suggestion
      if (PointerType *ItemType =
          dyn_cast<PointerType>(StructArgTy->getElementType(aggIdx))) {
```

Meinersbur reviewed (commented):
The LoadInst is loading a member of the struct whose alignment is not necessarily the same as the struct's alignment itself. for instance:

```c
struct {
  char a,b;
  int c;
} MyStruct;
```
`alignof(MyStruct)` should be 4, but `b` will offset 1 and only char-aligned. 

What the alignment of each member is, is determined by the ABI, and I am not sure it will always be the processor's preferred alignment. 

Consider adding alignment info to the store for live-out values as well.

DominikAdamski committed:
[OpenMP][CodeExtractor]Add align metadata to load instructions

LLVM IR language reference manual states that align metadata
tells the optimizer that the value loaded is known to be aligned
to a boundary specified by the integer value in the metadata node.

This information is used by the optimizer, for example,
to generate more efficient memcpy calls. The LLVM Optimizer requires
align metadata to generate optimized code because information about
the alignment of objects is lost during OpenMP target code generation
(outlining of loop body helper function).

DominikAdamski committed:
Revert "Applied remarks"

This reverts commit 3a41608e17f24df357e2d0131c353a182b659024.

DominikAdamski committed:
Revert "[CodeExtractor] Add align metadata to extracted pointers"

This reverts commit 802fef5f5e76dcdbaed87c1b665d74bce7c9c514.

DominikAdamski committed:
Merge branch 'main' into codeextractor_add_align_metadata

DominikAdamski commented:
Scope of changes:
1) Analyze MLIR map operations. Add information about the alignment of objects that are passed by reference to OpenMP GPU kernels.
2) Propagate alignment information to the outlined helper functions.

Meinersbur commented on llvm/lib/Transforms/Utils/CodeExtractor.cpp:
```diff
@@ -1612,8 +1625,31 @@ void CodeExtractor::emitFunctionBody(
+      LoadInst *LoadGEP =
+          new LoadInst(StructArgTy->getElementType(aggIdx), GEP,
+                       "loadgep_" + inputs[i]->getName(), newFuncRoot);
+      if (StructArgTy->getElementType(aggIdx)->isPointerTy()) {
```
Consider adding a code comment for what this new code does

Meinersbur commented on mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:
```diff
@@ -4437,9 +4442,21 @@ createDeviceArgumentAccessor(MapInfoData &mapData, llvm::Argument &arg,
+    // Add information about alignment of objects that are mapped by reference
+    if (v->getType()->isPointerTy() && alignmentValue) {
+      llvm::MDBuilder MDB(builder.getContext());
+      loadInst->setMetadata(
+          llvm::LLVMContext::MD_align,
+          llvm::MDNode::get(builder.getContext(),
+                            MDB.createConstant(llvm::ConstantInt::get(
+                                llvm::Type::getInt64Ty(builder.getContext()),
+                                alignmentValue))));
```
`CreateAlignedLoad` already receives alignment information. Why set `MD_align` with potentially different alignment?

Meinersbur reviewed (commented):
Consider also updatating the PR summary, you probably want to use it as commit message.

DominikAdamski committed:
Applied remarks

DominikAdamski commented on llvm/lib/Transforms/Utils/CodeExtractor.cpp:
```diff
@@ -248,6 +248,19 @@ buildExtractionBlockSet(ArrayRef<BasicBlock *> BBs, DominatorTree *DT,
   return Result;
 }
 
+static bool isAlignmentPreservedForAddrCast(const Triple &TargetTriple) {
```
Done

DominikAdamski commented on llvm/lib/Transforms/Utils/CodeExtractor.cpp:
```diff
@@ -1612,8 +1625,31 @@ void CodeExtractor::emitFunctionBody(
+        // alignment. Use pointers without casts if given target preserves
+        // alignment information for cast the operation.
+        if (isAlignmentPreservedForAddrCast(TargetTriple))
+          AlignmentValue =
+              inputs[i]->stripPointerCasts()->getPointerAlignment(DL).value();
```
The `getPointerAlignment` function is from the `Value` class. Do you think it's worth adding another version of this function to this basic class? If yes, I will create a separate patch that will contain a version of the `getPointerAlignment` function with two arguments. One of them will be a flag to get the value alignment without pointer casts.

Meinersbur reviewed (approved):
LGTM, thank you
